stl/inc/span

```diff
@@ -0,0 +1,679 @@
+        return _Extent;
+    }
+};
+
+template <>
+struct _Span_extent_type<dynamic_extent> {
+    constexpr _Span_extent_type() noexcept = default;
+
+    constexpr explicit _Span_extent_type(const size_t _Size) noexcept : _Mysize(_Size) {}
+
+    template <size_t _OtherExtent>
+    constexpr explicit _Span_extent_type(const _Span_extent_type<_OtherExtent> _Ext) noexcept : _Mysize(_Ext._Size()) {}
+
+    _NODISCARD constexpr size_t size() const noexcept {
+        return _Mysize;
+    }
+
+private:
+    size_t _Mysize{0};
+};
+
+// STRUCT TEMPLATE _Is_span
+template <class _Ty, size_t _Extent>
+class span;
+
+template <class>
+struct _Is_span : false_type {};
+
+template <class _Ty, size_t _Extent>
+struct _Is_span<span<_Ty, _Extent>> : true_type {};
+
+// STRUCT TEMPLATE _Is_std_array
+template <class>
+struct _Is_std_array : false_type {};
+
+template <class _Ty, size_t _Extent>
+struct _Is_std_array<array<_Ty, _Extent>> : true_type {};
+
+// STRUCT TEMPLATE _Is_span_convertible_range
+template <class _Cont, class _Ty>
+struct _Is_span_convertible_range
+    : is_convertible<remove_pointer_t<decltype(_STD data(_STD declval<_Cont&>()))> (*)[], _Ty (*)[]>::type {};
+
+// STRUCT TEMPLATE _Has_container_interface
+template <class, class = void>
+struct _Has_container_interface : false_type {};
+
+template <class _Cont>
+struct _Has_container_interface<_Cont,
+    void_t<decltype(_STD data(_STD declval<_Cont&>())), decltype(_STD size(_STD declval<_Cont&>()))>> : true_type {};
+
+// STRUCT TEMPLATE _Is_span_compatible_container
+// clang-format off
+template <class _Cont, class _Ty>
+struct _Is_span_compatible_container : conjunction<
+    negation<is_array<_Cont>>,
+    negation<_Is_span<remove_const_t<_Cont>>>,
+    negation<_Is_std_array<remove_const_t<_Cont>>>,
+    _Has_container_interface<_Cont>,
+    _Is_span_convertible_range<_Cont, _Ty>>::type {};
+// clang-format on
+
+// [views.span]
+// CLASS TEMPLATE span
+template <class _Ty, size_t _Extent = dynamic_extent>
+class span : public _Span_extent_type<_Extent> {
+public:
+    using _Mybase                = _Span_extent_type<_Extent>;
+    using element_type           = _Ty;
+    using value_type             = remove_cv_t<_Ty>;
+    using index_type             = size_t;
+    using difference_type        = ptrdiff_t;
+    using pointer                = _Ty*;
+    using const_pointer          = const _Ty*;
+    using reference              = _Ty&;
+    using const_reference        = const _Ty&;
+    using iterator               = _Span_iterator<_Ty>;
+    using const_iterator         = _Span_iterator<const _Ty>;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;
+
+    static constexpr index_type extent = _Extent;
+
+    // [span.cons] Constructors, copy, and assignment
+    template <size_t _Ext = _Extent, enable_if_t<_Ext == 0 || _Ext == dynamic_extent, int> = 0>

```

Comment:
Not sure whether it's worth commenting here: `enable_if` (rather than `requires`) works even though this is a special member here because the other 2 overloads suppress generation of a compiler generated default ctor.

Reply:
*Any* user-declared constructor suppresses generation of the default constructor ([[class.default.ctor]/1](http://eel.is/c++draft/class.default.ctor#1)).

Reply:
The problem is that we need to conditionally enable it. Otherwise we would have to specialize `_Span_Extent` to prohibit the default contructor for `Extent != 0`

Btw may I comment that conceptually splitting static spans into two categories feels like a bad desing for no real value?